### PR TITLE
Generalize ShapeInfo::reinit

### PR DESCRIPTION
**Pull request by peterrum**
Depends on #11361.

### CODE REVIEW

File: include/deal.II/matrix_free/shape_info.templates.h [hunk 1]
```diff
@@ -316,54 +316,91 @@ namespace internal
...
-                  .n_faces();
+              AssertThrow(reference_cell_type == temp.first,
+                          ExcNotImplemented());
+
+              this->n_q_points_face = quad_face[0].size();
+
+              // TODO: lift limitation
+              for (unsigned int i = 1; i < quad_face.size(); ++i)
+                AssertThrow(this->n_q_points_face == quad_face[i].size(),
+                            ExcNotImplemented());
```
Comment (peterrum) [inline include/deal.II/matrix_free/shape_info.templates.h:332]: This is the place where some (minor) changes have to be done for wedges and pyramids. I'll address this once #11361 and this PR are meged.

**Comment by peterrum** [comment]
I have rebased this PR and have integrated some additional features needed by #11401. 

@kronbichler #11387, #11400, and #11401 heavily depend on each other and on a design decision. We should talk about these in the new year!

File: include/deal.II/matrix_free/util.h [hunk 2]
```diff
@@ -45,6 +49,55 @@ namespace internal
...
     }
 
+    template <int dim>
+    inline std::pair<ReferenceCell::Type, dealii::hp::QCollection<dim - 1>>
+    get_face_quadrature_collection(const Quadrature<dim> &quad,
+                                   const bool             do_assert = true)
+    {
+      if (dim == 2 || dim == 3)
+        for (unsigned int i = 1; i <= 3; ++i)
+          if (quad == Simplex::QGauss<dim>(i))
```
Comment (kronbichler) [inline include/deal.II/matrix_free/util.h:59]: Does this need to be adjusted now that we have more simplex quadrature formulas?

**Review by kronbichler — APPROVED** [review]
Looks good, if we can have a good answer to the question of general quadrature formulas.

File: include/deal.II/matrix_free/util.h (hunk 2 above)
Comment (peterrum) [reply include/deal.II/matrix_free/util.h:59]: Indeed it needs to be updated. Generally, I am not happy by the approach here but I don't have a better idea (that is too intrusive - making the face quadrature rule part of `MatrixFree::reinit()` just like in the case of `FEValues`). But, nevertheless let's proceed on this front and think resolve this issue till the release!

File: include/deal.II/matrix_free/util.h [hunk 3]
```diff
@@ -45,6 +49,70 @@ namespace internal
...
+                          dealii::hp::QCollection<dim - 1>(tri, tri, tri)};
+                else
+                  return {ReferenceCells::Tetrahedron,
+                          dealii::hp::QCollection<dim - 1>(tri, tri, tri, tri)};
+              }
+
+          for (unsigned int i = 1; i <= 5; ++i)
+            if (quad == Simplex::QGauss<dim>(i))
+              {
+                Simplex::QWitherdenVincent<dim - 1> tri(i);
```
Comment (peterrum) [inline include/deal.II/matrix_free/util.h:75]: FYI @drwells :)

**Review by kronbichler — APPROVED** [review]

**Pull request merged**